app/assets/javascripts/components/story/CollapsedStory/index.js

```diff
@@ -25,65 +29,187 @@ const storyClassName = (story, additionalClassname = '') => {
+
+const checkStoryPosition = (ref, dragIndex, hoverIndex, monitor) => {
+  const hoverBoundingRect = ref.current.getBoundingClientRect()
+  const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2
+  const clientOffset = monitor.getClientOffset()
+  const hoverClientY = clientOffset.y - hoverBoundingRect.top
+
+  return isDraggedInTheSamePlaceOfHolder(dragIndex, hoverIndex)
+    || canContinue(dragIndex, hoverIndex, hoverClientY, hoverMiddleY)
+}
+
+const canChangePosition = (ref, monitor, storySource, hoverIndex, newPosition) => {
+  if(!ref.current) {
+    return true
+  }
+
+  if(storySource.position === newPosition) {
+    return true
+  }
+
+  if (storySource.index === hoverIndex) {
+    return true
+  }
+
+  if(checkStoryPosition(ref, storySource.index, hoverIndex, monitor)) {
+    return true
+  }
+}
+
+const canChangeColumn = (state, column) =>
+  (state === 'unscheduled' || state === 'unstarted') && column !== 'done';
+
+const isADropTarget = (monitor) => !monitor.getDropResult()
+
+export const CollapsedStory = ({
+  onToggle,
+  story,
+  dragDropStory,
+  project,
+  className,
+  title,
+  index,
+  stories,
+  column,
+  updateCollapsedStory,
+  highlight,
+  isHighlightable
+}) => {
+  const ref = useRef(null)
+  const [, drop] = useDrop({
+    accept: type.story,
+    canDrop() {
+      return column !== 'done'
+    },
+    hover(item, monitor) {
+      const storySource = monitor.getItem();
+      const hoverIndex = index
+      const storiesArray = storySource.stories;
+
+      if (storySource.column !== column){
+        return true
+      };
+
+      const newPosition = getNewPosition(item, story, storiesArray, index);
+
+      if (canChangePosition(ref, monitor, storySource, hoverIndex, newPosition)) {
+        return true;
+      }
+
+      dragDropStory(storySource.id, storySource.projectId, {
+        position: newPosition
+      })
     }
-  </div>
+  })
+  const [{ isDragging }, drag] = useDrag({
+    item: { type: type.story, ...story, index, stories, column},
+    collect: monitor => ({
+      isDragging: monitor.isDragging(),
+    }),
+    canDrag() {
+      return story.state === "accepted" ? false : true
+    },
+    begin() {
+      return { ...story, index, stories, column }
+    },
+    end(item, monitor) {
+      if(isADropTarget(monitor)) return null;
+      const {state, column} = monitor.getDropResult();
+
+      if (canChangeColumn(state, column) && !Story.isUnestimatedFeature(story)) {
+        dragDropStory(item.id, item.projectId, {
+          state: column === 'backlog' ? 'unstarted' : 'unscheduled'
+        });
+      }
+    }
+  })
+  drag(drop(ref))
+  return (
+    <div
+      ref={ref}
+      className={storyClassName(story, className)}
+      onClick={onToggle}
+      title={title}
+      style={{
+        opacity: isDragging ? 0 : 1
```

Comment:
What you think about use css classes here following the BEM pattern instead inject the inline style? I believe than adding more one conditional on `storyClassName` with new classes works. And the name of classes can be `Story--dragging` with opacity 0.